pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -464,53 +468,86 @@ func (gce *GCECloud) EnsureLoadBalancer(name, region string, requestedIP net.IP,
 	// function in order to maintain the invariant that "if the forwarding rule
 	// exists, the LB has been fully created".
 	ipAddress := ""
+
 	// Through this process we try to keep track of whether it is safe to
 	// release the IP that was allocated.  If the user specifically asked for
 	// an IP, we assume they are managing it themselves.  Otherwise, we will
 	// release the IP in case of early-terminating failure or upon successful
 	// creating of the LB.
-	releaseStaticIP := false
+	isUserOwnedIP := false // if this is set, we never release the IP
+	isSafeToReleaseIP := false
+	defer func() {
+		if isUserOwnedIP {
+			return
+		}
+		if isSafeToReleaseIP {
+			if err := gce.deleteStaticIP(name, region); err != nil {
+				glog.Errorf("failed to release static IP %s for load balancer (%v, %v): %v", ipAddress, name, region, err)
+			}
+			glog.V(2).Infof("EnsureLoadBalancer(%v): released static IP %s", name, ipAddress)
+		} else {
+			glog.Warningf("orphaning static IP %s during update of load balancer (%v, %v): %v", ipAddress, name, region, err)
+		}
+	}()
+
 	if requestedIP != nil {
 		// If a specific IP address has been requested, we have to respect the
 		// user's request and use that IP. If the forwarding rule was already using
 		// a different IP, it will be harmlessly abandoned because it was only an
 		// ephemeral IP (or it was a different static IP owned by the user, in which
 		// case we shouldn't delete it anyway).
-		if err := gce.projectOwnsStaticIP(name, region, requestedIP.String()); err != nil {
-			return nil, err
+		if isStatic, err := gce.projectOwnsStaticIP(name, region, requestedIP.String()); err != nil {
```

Comment:
I have another crazy situation for you:
1. Alice creates LB without a requested IP
2. Alice takes the IP that got assigned to the LB and puts it in the requestedIP field (which appears to be supported, given all the code being added)
3. Alice updates the service in a way that requires deleting and recreating the forwarding rule
4. The first time through this code, we promote the IP to static in the `if requestedIP.String() == fwdRuleIP` condition
5. We delete the fwd rule but fail to recreate it, causing the IP to be kept around
6. On retry, this isStatic check evaluates to true because the requested IP was temporarily promoted to static last time around, causing us to think it's a user-owned IP and not release it when we're done.

This is such a minor case that it may not be worth handling at this point given the more urgent issues, but it's at least worth making note of somewhere.